### ISSUE TITLE
emmeans for a MCMCglmm object: Error in .find.by.rows(object@grid, "trait"), in the case of a binary logit model

### ISSUE THREAD

**Issue by atrolle** [rvlenth/emmeans#149]
Hi Russell,

I'm now using the latest GitHub version of emmeans for plotting marginal effects for a binary logit model.
To build that model, I used `MCMCglmm`, again.
Here is the command:
`model.mcmc <- MCMCglmm(yvar ~ Gender + Module + Wish,
                                    data = data, family = "categorical", prior = priors,
                                         burnin = 20000, nitt = 40000, thin = 10, pr=FALSE, pl=TRUE)
`

`yvar` is the binary response (a factor with 2 levels); `Wish` is a factor with 2 levels, `Gender` a factor with 2 levels and `Module` a factor with 8 levels.
To write this command, I followed Jarrod Hadfield's course notes.

But, when creating the grid with:
`grid.emm <- ref_grid(model.mcmc, data = data, type = "response", mode = "multinomial")`

I have now the following error:

> Error in .find.by.rows(object@grid, "trait") : 
>   'by' variables are not all in the grid

Again, could you enlighten me on this issue?

Thank you,
Arnaud.

**Comment by rvlenth** [comment]
I don’t know. Is this something that worked with the latest CRAN version?

Sent from my iPhone

On Oct 18, 2019, at 1:22 AM, atrolle <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi Russell,

I'm now using the latest GitHub version of emmeans for plotting marginal effects for a binary logit model.
To build that model, I used MCMCglmm, again.
Here is the command:
model.mcmc <- MCMCglmm(yvar ~ Gender + Module + Wish, data = data, family = "categorical", prior = priors, burnin = 20000, nitt = 40000, thin = 10, pr=FALSE, pl=TRUE)

yvar is the binary response (a factor with 2 levels); Wish is a factor with 2 levels, Gender a factor with 2 levels and Module a factor with 8 levels.
To write this command, I followed Jarrod Hadfield's course notes.

But, when creating the grid with:
grid.emm <- ref_grid(model.mcmc, data = data, type = "response", mode = "multinomial")

I have now the following error:

Error in .find.by.rows(object@grid, "trait") :
'by' variables are not all in the grid

Again, could you enlighten me on this issue?

Thank you,
Arnaud.

**Comment by atrolle** [comment]
No, It didn't work with the latest CRAN version as well.
There may be a problem with the MCMCglmm definition of a binary logistic model, for which `trait` is not used (since not necessary), whereas it is used for the MCMCglmm definition of a multinomial logistic model (with more than 2 categories for the response).

**Comment by rvlenth** [comment]
If it isn't multinomial, then you shouldn't specify `mode = "multinomial"`. Maybe it'll work if you take that out?

In general, I have had considerable difficulty supporting MCMCglmm because it is complex, keeps changing, and the developer doesn't answer my questions. So I'm afraid I can't promise a lot.

**Comment by atrolle** [comment]
If I specify: 
`grid.emm <- ref_grid(model.mcmc, data = data, type = "response")`
it works, but the values in `grid.emm@post.beta` are on the logit scale, not on the probability scale.

**Comment by atrolle** [comment]
Maybe your `ref_grid` function could be extended to the "binomial" case for MCMCglmm models, such as follows:
`grid.emm <- ref_grid(model.mcmc, data = data, type = "response", mode = "binomial")`

An input in the code of your `emm_basis.MCMCglmm` function would be required:
```
emm_basis.MCMCglmm = function(object, trms, xlev, grid, vcov., 
                              mode = c("default", "multinomial", "binomial"), misc, ...) {
    nobs.MCMCglmm = function(object, ...) 1   # prevents warning about nobs
    m = model.frame(trms, grid, na.action = na.pass, xlev = xlev)
    X = model.matrix(trms, m, contrasts.arg = NULL)
    Sol = as.matrix(object$Sol)[, seq_len(object$Fixed$nfl)] # toss out random effects if included
    bhat = apply(Sol, 2, mean)
    if (missing(vcov.))
        V = cov(Sol)
    else
        V = .my.vcov(object, vcov.)
    if (is.null(misc))
        misc = list()
    mode = match.arg(mode)
    if (mode == "multinomial") {
        misc$postGridHook = .MCMCglmm.multinom.postGrid
    } else {
       if(mode == "binomial") {
          misc$postGridHook = .MCMCglmm.binom.postGrid
       }
    }
    list(X = X, bhat = bhat, nbasis = matrix(NA), V = V, 
         dffun = function(k, dfargs) Inf, dfargs = list(), 
         misc = misc, post.beta = Sol)
}
```

The new local function `.MCMCglmm.binom.postGrid` would look like:
```
.MCMCglmm.binom.postGrid = function(object) {
    linfct = object@linfct
    misc = object@misc
    post.lp = object@post.beta %*% t(linfct)
    sel = .find.by.rows(object@grid, NULL)
    k = length(sel)
    cols = unlist(sel)
    scal = sqrt(1 + 2 * (16 * sqrt(3) / (15 * pi))^2 / (k + 1))  # scaling const for logistic
    # I'm assuming here that diag(IJ) = 2 / (k + 1)
    object@post.beta = post.p = t(apply(post.lp, 1, function(l) {
        expX = exp(cbind(0, matrix(l[cols], ncol = k)) / scal)
        as.numeric(apply(expX, 1, function(z) z / sum(z)))
    })) # These results come out with response levels varying the fastest.
   object@post.beta = object@post.beta[,seq(2,ncol(object@post.beta),2)]
    post.p = post.p[,seq(2,ncol(post.p),2)]
    object@bhat = apply(post.p, 2, mean)
    object@V = cov(post.p)
    preds = object@roles$predictors
 ##   object@roles$predictors  = preds[preds != "trait"]
 ##   object@levels[["trait"]] = NULL
 ##   object@levels = c(list(misc$resp.levs), object@levels)
 ##   names(object@levels)[1] = misc$trait
    object@grid = do.call(expand.grid, object@levels)
    
    misc$postGridHook = misc$tran = misc$inv.lbl = 
        misc$trait = misc$resp.levs = NULL
    misc$display = object@model.info$nesting = NULL
    misc$estName = "prob"
    object@linfct = diag(1, ncol(post.p))
    object@misc = misc
    
    object
}
```

I've badly coded the thing, but it's a start step.
Do you think it possible (and useful to other users) to include such an extension in `emmeans` in the next future?

**Comment by rvlenth** [comment]
If I understand the first of your last two comments, it already works, or almost works, in its present form. 

First of all, a `type` argument if `ref_grid` only affects *how it will be summarized later*. It does not affect what's stored in the `post.beta` slot. So it is *not an error* if the `post.beta` slot is on the logit scale. Please try:
```r
grid.emm <- ref_grid(model.mcmc, data = data, type = "response")
str(grid.emm)
```
If you see the logit transformation noted there, then everything is exactly as needed, and if you do
```r
summary(grid.emm)
```
it will summarize the grid on the response scale. In other words, *by design*, the above summarization is equivalent to:
```r
rg <-  ref_grid(model.mcmc, data = data)
summary(rg, type = "response")
```

Second, if you use `transform` instead of `type`:
```r
grid.emm.r <- ref_grid(model.mcmc, data = data, transform = "response")
## or equivalently
##    grid.emm.r <- regrid(ref_grid((model.mcmc, data = data))
```
this *will* put the `post.beta` slot on the response scale. See the documentation for `ref_grid` and the discussion therein on the distinction between `type` and `transform`.

If the `str()` results don't show the logit transformation, then I think the only thing that is needed is code to detect the transformation; and a workaround for that in the interim is
```r
grid.emm <- update(ref_grid(model.mcmc, data = data, type = "response"), tran = "logit")
summary(grid.emm)
```

In summary, the design of **emmeans** for other generalized linear models, Bayesian or not, is to keep the predictions on the link scale unless explicitly converted to the response scale via `regrid()` or a `transform` argument. I would not want to have `MCMCglmm` objects handled differently.

**Comment by atrolle** [comment]
> First of all, a type argument if ref_grid only affects how it will be summarized later. It does not affect what's stored in the post.beta slot. So it is not an error if the post.beta slot is on the logit scale.

Ok. Sorry, I didn't mean it was an error.

> Second, if you use `transform` instead of `type`:
> ```grid.emm.r <- ref_grid(model.mcmc, data = data, transform = "response")
> ## or equivalently
> ##    grid.emm.r <- regrid(ref_grid((model.mcmc, data = data))
> ```
> this will put the `post.beta` slot on the response scale.

Actually, it doesn't show the results back-transformed to the probability scale, but still the results on the logit scale.

> If the str() results don't show the logit transformation, then I think the only thing that is needed is code to detect the transformation; and a workaround for that in the interim is
> ```
> grid.emm <- update(ref_grid(model.mcmc, data = data, type = "response"), tran = "logit")
> summary(grid.emm)
> ``` 

This workaround works fine. I'm going to use it. Thank you for this solution.

**Comment by rvlenth** [comment]
I have added rudimentary support for detecting the link. This is based on the `"family"` slot in the returned object, which is messy because (1) it is not necessarily unique, and (2) it doesn't explicitly give the link function used - a poor design choice, IMHO. Not all families are supported, by far. I hope I'm guessing it correctly for the families that are. Here's an example, `m2a.5` in the course notes. It's a Poisson model, but it should work similarly for your binomial model.

```r
> rg = ref_grid(m2a.5, data = Traffic)
> rg
'emmGrid' object with variables:
    limit = no, yes
    year = 1961, 1962
    day = 46.5
Transformation: “log” 
```
So we see it detects the link. Now, some summaries:
```r
> summary(rg)   # on the link scale
 limit year  day prediction lower.HPD upper.HPD
 no    1961 46.5       3.11      3.03      3.19
 yes   1961 46.5       2.94      2.82      3.06
 no    1962 46.5       3.05      2.94      3.14
 yes   1962 46.5       2.87      2.77      2.98

Point estimate displayed: median 
Results are given on the log (not the response) scale. 
HPD interval probability: 0.95 

> summary(rg, type = "r")   # on the response scale
 limit year  day response lower.HPD upper.HPD
 no    1961 46.5     22.5      20.6      24.4
 yes   1961 46.5     18.9      16.4      21.0
 no    1962 46.5     21.0      18.9      23.0
 yes   1962 46.5     17.7      15.9      19.7

Point estimate displayed: median 
Results are back-transformed from the log scale 
HPD interval probability: 0.95 
```
A subtlety of MCMCglmm is it always fits an over-dispersion model. So in his course notes, you often see him do a fancy correction when back-transforming -- see (2.11) and (2.12) for example. This correction compensates for bias when back-transforming in the presence of random effects. To do this, we need the estimated random effects:
```r
> head(m2a.5$VCV)
Markov Chain Monte Carlo (MCMC) output:
Start = 3001 
End = 3061 
Thinning interval = 10 
          units
[1,] 0.08921038
[2,] 0.11199005
[3,] 0.08900765
[4,] 0.09603395
[5,] 0.10671439
[6,] 0.10706271
[7,] 0.09151697
```
(If there is more than one variance involved, you need to add up the appropriate pieces, again per his examples.) So a less biased version of the back-transformed results are obtained as follows:
```r
> summary(rg, type = "r", bias.adj = TRUE, sigma = sqrt(m2a.5$VCV[,"units"]))
 limit year  day response lower.HPD upper.HPD
 no    1961 46.5     23.6      21.6      25.5
 yes   1961 46.5     19.9      17.5      22.2
 no    1962 46.5     22.1      19.7      24.2
 yes   1962 46.5     18.6      16.7      20.7

Point estimate displayed: median 
Results are back-transformed from the log scale 
Bias adjustment applied based on sigma = (various values) 
HPD interval probability: 0.95 
```
I'll push the update shortly. I'm closing the issue, but you can still comment and I can reopen if needed.

**Comment by rvlenth** [comment]
Note: Please see Issue #253 relating to multinomial models. A new version 1.5.4 has been submitted to CRAN that corrects this error
